lib/libc/sys/flock.2
.\"	BSDI flock.2,v 2.2 1995/07/27 17:56:23 bostic Exp
.\"
.\" Copyright (c) 1983, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)flock.2	8.2 (Berkeley) 12/11/93
.\"
.Dd December 11, 1993
.Dt FLOCK 2
.Os BSD 4.2
.Sh NAME
.Nm flock
.Nd "apply or remove an advisory lock on an open file"
.Sh SYNOPSIS
.Fd #include <fcntl.h>
.Ft int
.Fn flock "int fd" "int operation"
.Sh DESCRIPTION
.Fn Flock
applies or removes an
.Em advisory
lock on the file associated with the file descriptor
.Fa fd .
The
.Fa operation
parameter is created from the
.Em or'd
operation bit masks defined in
.Aq Pa fcntl.h :
.Pp
.Bd -literal -offset indent -compact
#define LOCK_SH	0x01	/* shared file lock */
#define LOCK_EX	0x02	/* exclusive file lock */
#define LOCK_NB	0x04	/* don't block when locking */
#define LOCK_UN	0x08	/* unlock file */
.Ed
.Pp
A lock is applied by specifying either
.Dv LOCK_SH
or
.Dv LOCK_EX
with the optional inclusion of
.Dv LOCK_NB .
To unlock an existing lock,
.Dv operation
should be
.Dv LOCK_UN .
.Pp
Advisory locks allow cooperating processes to perform
consistent operations on files, but do not guarantee
consistency (i.e., processes may still access files
without using advisory locks possibly resulting in
inconsistencies).
.Pp
The locking mechanism allows two types of locks:
.Em shared
locks and
.Em exclusive
locks.
At any time multiple shared locks may be applied to a file,
but at no time are multiple exclusive, or both shared and exclusive,
locks allowed simultaneously on a file.  
.Pp
A shared lock may be
.Em upgraded
to an exclusive lock, and vice versa, simply by specifying
the appropriate lock type; this results in the previous
lock being released and the new lock applied (possibly
after other processes have gained and released the lock).
.Pp
Requesting a lock on an object that is already locked
normally causes the caller to be blocked until the lock may be
acquired.  If
.Dv LOCK_NB
is included in
.Fa operation ,
then this will not happen; instead the call will fail and
the error
.Er EWOULDBLOCK
will be returned.
.Sh NOTES
Locks are on files, not file descriptors.  That is, file descriptors
duplicated through
.Xr dup 2
or
.Xr fork 2
do not result in multiple instances of a lock, but rather multiple
references to a single lock.  If a process holding a lock on a file
forks and the child explicitly unlocks the file, the parent will
lose its lock.
.Pp
Processes blocked awaiting a lock may be awakened by signals.
.Sh RETURN VALUES
Zero is returned if the operation was successful;
on an error a -1 is returned and an error code is left in
the global location
.Va errno .
.Sh ERRORS
The
.Fn flock
call fails if:
.Bl -tag -width EWOULDBLOCKAA
.It Bq Er EWOULDBLOCK
The file is locked and the
.Dv LOCK_NB
option was specified.
.It Bq Er EBADF
The argument
.Fa fd
is an invalid descriptor.
.It Bq Er EINVAL
The argument
.Fa fd
refers to an object other than a file.
.El
.Sh SEE ALSO
.Xr open 2 ,
.Xr close 2 ,
.Xr dup 2 ,
.Xr execve 2 ,
.Xr fork 2
.Sh HISTORY
The
.Nm
function call appeared in
.Bx 4.2 .


core_contrib/qpopper/common/flock.c
/*
 * Copyright (c) 2003 QUALCOMM Incorporated.  All rights reserved.
 * The file License.txt specifies the terms for use, modification,
 * and redistribution.
 */


/*
 * Copyright (c) 1989 Mark Davies
 * Copyright (c) 1990 Andy Linton
 * Copyright (c) 1990 Victoria University of Wellington.
 *
 * Redistribution and use in source and binary forms are permitted provided
 * that: (1) source distributions retain this entire copyright notice and
 * comment, and (2) distributions including binaries display the following
 * acknowledgement:  ``This product includes software developed by the
 * Victoria University of Wellington, New Zealand and its contributors''
 * in the documentation or other materials provided with the distribution
 * and in all advertising materials mentioning features or use of this
 * software.
 * Neither the name of the University nor the names of its contributors may
 * be used to endorse or promote products derived from this software without
 * specific prior written permission.
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 */


#include "config.h"

#ifndef HAVE_FLOCK /*  defined(SYSV) || defined(ISC) */

#include <unistd.h>
#include <sys/types.h>
#include <errno.h>
#ifdef ISC
#  include <net/errno.h>
#endif
#include <flock.h>               /* defines normally in <sys/file.h> */


#if defined(F_SETLK) && defined(F_SETLKW)

int
flock(fd, operation)
int fd, operation;
{
    int op, ret;
    struct flock arg;
    extern int errno;

    op = (LOCK_NB & operation) ? F_SETLK : F_SETLKW;
    
    arg.l_type = (LOCK_EX & operation) ? F_WRLCK :
        (LOCK_SH & operation) ? F_RDLCK : F_UNLCK;
    arg.l_whence = arg.l_start = arg.l_len = 0;
    
    if ((ret = fcntl(fd, op, &arg)) == -1) {
        if (errno == EACCES || errno == EAGAIN)
            errno = EWOULDBLOCK;
    }
    return (ret);
}

#else

int
flock(fd, operation)
int fd, operation;
{
    int op, ret;
    extern int errno;

    op = (LOCK_UN & operation) ? F_ULOCK :
      (LOCK_NB & operation) ? F_TLOCK : F_LOCK;

    if ((ret = lockf(fd, op, 0)) == -1) {
        if (errno == EACCES || errno == EAGAIN)
            errno = EWOULDBLOCK;
    }

    return (ret);
}

#endif /* F_SETLK && F_SETLKW */

#endif /* HAVE_FLOCK */

